samples/custom_dispatch/cpu/mlp_plugin/mlp_spec.mlir

```diff
@@ -2,51 +2,29 @@
 // an implementation implemented by a system plugin.
 // Is used along with samples/custom_dispatch/cpu/plugin/mlp.mlir
 
-#x86_64_target = #hal.executable.target<"llvm-cpu", "embedded-elf-x86_64", {
-  data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128",
-  native_vector_size = 32 : index,
-  target_triple = "x86_64-none-elf"
-}>
-
-#cpu_target = #hal.device.target<"llvm-cpu", {
-  executable_targets = [
-    #x86_64_target
-  ]
-}>
-
 module attributes {transform.with_named_sequence} {
 
   // Executable that stages call to the external functions.
-  hal.executable private @executable {
-    hal.executable.variant public @x86_64 target(#x86_64_target) {
-      hal.executable.export public @mlp ordinal(0)
-          layout(#hal.pipeline.layout<push_constants = 3, sets = [
-            <0, bindings = [
-              <0, storage_buffer, ReadOnly>,
-              <1, storage_buffer, ReadOnly>,
-              <2, storage_buffer>
-            ]>
-          ]>) {
-      ^bb0(%device : !hal.device):
-        %c1 = arith.constant 1 : index
-        hal.return %c1, %c1, %c1 : index, index, index
-      }
-      builtin.module {
-        func.func private @mlp_external(%lhs : memref<?x?xf32>, %rhs : memref<?x?xf32>, %result : memref<?x?xf32>, %m : i32, %n : i32, %k : i32)
-        func.func @mlp() {
-          %m_i32 = hal.interface.constant.load[0] : i32
-          %n_i32 = hal.interface.constant.load[1] : i32
-          %k_i32 = hal.interface.constant.load[2] : i32
-          %c0 = arith.constant 0 : index
-          %m = arith.index_cast %m_i32 : i32 to index
-          %n = arith.index_cast %n_i32 : i32 to index
-          %k = arith.index_cast %k_i32 : i32 to index
-          %lhs = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c0) : memref<?x?xf32>{%m, %k}
-          %rhs = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c0) : memref<?x?xf32>{%k, %n}
-          %result = hal.interface.binding.subspan set(0) binding(2) type(storage_buffer) alignment(64) offset(%c0) : memref<?x?xf32>{%m, %n}
-          func.call @mlp_external(%lhs, %rhs, %result, %m_i32, %n_i32, %k_i32) : (memref<?x?xf32>, memref<?x?xf32>, memref<?x?xf32>, i32, i32, i32) -> ()
-          return
-        }
+  stream.executable private @executable {
+    stream.executable.export public @mlp workgroups() -> (index, index, index) {
+      %c1 = arith.constant 1 : index
+      hal.return %c1, %c1, %c1 : index, index, index
+    }
+    builtin.module {
+      func.func private @mlp_external(%lhs : memref<f32>, %lhs_offset : index, %rhs : memref<f32>, %rhs_offset : index, %result : memref<f32>, %result_offset : index, %m : i32, %n : i32, %k : i32) attributes {llvm.bareptr}
+      func.func @mlp(%arg0: !stream.binding, %arg1: !stream.binding, %arg2: !stream.binding, %arg3: i32, %arg4: i32, %arg5 : i32) {
```

Comment:
as mentioned above the args can be index here (no need for the casts then and both this and your pdl will be smaller/have less noise)